packages/preferences/src/browser/views/preference-editor-widget.ts

```diff
@@ -0,0 +1,316 @@
+
+    protected scrollContainer: HTMLDivElement;
+    /**
+     * Guards against scroll events and selection events looping into each other. Set before this widget initiates a selection.
+     */
+    protected currentModelSelectionId = '';
+    /**
+     * Permits the user to expand multiple nodes without each one being collapsed on a new selection.
+     */
+    protected lastUserSelection = '';
+    protected isAtScrollTop = true;
+    protected firstVisibleChildID = '';
+    protected renderers = new Map<string, GeneralPreferenceNodeRenderer>();
+    // The commonly used section will duplicate preference ID's, so we'll keep a separate list of them.
+    protected commonlyUsedRenderers = new Map<string, GeneralPreferenceNodeRenderer>();
+
+    @inject(PreferenceService) protected readonly preferenceService: PreferenceService;
+    @inject(PreferenceTreeModel) protected readonly model: PreferenceTreeModel;
+    @inject(PreferenceNodeRendererFactory) protected readonly rendererFactory: PreferenceNodeRendererFactory;
+    @inject(PreferencesScopeTabBar) protected readonly tabbar: PreferencesScopeTabBar;
+
+    @postConstruct()
+    protected init(): void {
+        this.id = PreferencesEditorWidget.ID;
+        this.title.label = PreferencesEditorWidget.LABEL;
+        this.addClass('settings-main');
+        this.toDispose.pushAll([
+            this.preferenceService.onPreferencesChanged(e => this.handlePreferenceChanges(e)),
+            this.model.onFilterChanged(e => this.handleFilterChange(e)),
+            this.model.onSelectionChanged(e => this.handleSelectionChange(e)),
+        ]);
+        this.createContainers();
+    }
+
+    protected createContainers(): void {
+        const innerWrapper = document.createElement('div');
+        innerWrapper.classList.add('settings-main-scroll-container');
+        this.scrollContainer = innerWrapper;
+        innerWrapper.addEventListener('scroll', this.onScroll, { passive: true });
+        this.node.appendChild(innerWrapper);
+        const noLeavesMessage = document.createElement('div');
+        noLeavesMessage.classList.add('settings-no-results-announcement');
+        noLeavesMessage.textContent = 'That search query has returned no results.';
+        this.node.appendChild(noLeavesMessage);
+    }
+
+    protected handleFilterChange(e: PreferenceFilterChangeEvent): void {
+        const { isFiltered } = this.model;
+        const currentFirstVisible = this.firstVisibleChildID;
+        if (e.source === PreferenceFilterChangeSource.Search) {
+            this.handleSearchChange(isFiltered);
+        } else if (e.source === PreferenceFilterChangeSource.Scope) {
+            this.handleScopeChange(isFiltered);
+        } else if (e.source === PreferenceFilterChangeSource.Schema) {
+            this.handleSchemaChange(isFiltered);
+        } else {
+            ((otherSource: never) => { throw new Error('You should handle all possible sources of change!'); })(e.source);
```

Comment:
What's happening here?

Reply:
This is a way to get to TS to track whether there are any unhandled cases. Since TS knows that it's dealing with an enum, when it knows that all options have been exhausted, it retypes `e.source` as `never`. If it hasn't done that retyping by the time it reaches this branch - i.e. if you haven't handled all variants of the enum - then 'calling' this function with a non-never type is a build-time error. I've adapted this from a suggestion from Akos elsewhere and this [SO](https://stackoverflow.com/a/39419171) answer.

Reply:
Interesting read, thanks!